### ISSUE TITLE
Update v8 version

### ISSUE THREAD

**Issue by rahulchaphalkar** [envoyproxy/envoy#28336]
*Title*: *Updating v8 requires uploading static snapshot at `https://storage.googleapis.com/envoyproxy-wee8/v8-{version}.tar.gz` ?*

*Description*:
The only way to update v8 is to have the relevant version's snapshot at the above bucket, is that right? 
I was hoping to update to v8's main branch until the next release is available, as [discussed here](https://github.com/envoyproxy/envoy/pull/27890#discussion_r1255210635), but looks like that is not possible.

**Comment by phlax** [comment]
>  The only way to update v8 is to have the relevant version's snapshot at the above bucket, is that right?

i think yes - its not an ideal situation - but i think that is how it is currently setup

**Comment by mpwarres** [comment]
I can upload a new static snapshot, but would like to get clarification: is the proposal that we have Envoy use a version of v8 that is not an official release? How long would it be until the upstream v8 change is available in a standard release?

**Comment by mpwarres** [comment]
@rahulchaphalkar what is the specific version tag of v8 that you need the dependency updated to?

**Comment by rahulchaphalkar** [comment]
Thanks for the details.
@mpwarres I was hoping to use v8 commit `cc49dce9f5cb14daee8cc0946e996ba999f97d29` till a newer release of v8 is available, at which point we can switch to that release.
Looking at `repository_locations.bzl` and `wee8-fetch-deps.sh`, it seems perhaps other dependencies like `base_trace_event`, `icu`, `zlib` etc. might need to be updated in `repository_locations.bzl` if v8 is updated? I can do that if required.

**Comment by PiotrSikora** [comment]
Using a random commit from the mainline branch of V8 in a production software in order to add support for a niche and bleeding edge feature is definitely something that I'd object to in principle (but that's @mpwarres's role now).

Even more so, because V8 is not safe / stable on their main branch, which is used for active development with a lot reverts and no security updates... Even the commit you referenced was originally reverted and only relanded last week.

Having said that, it looks that it's a tiny change, so if there are no other code dependencies, then I'd recommend cherry-picking that change and carrying it in `bazel/v8.patch` until it ships in a stable release of V8.

Also, it looks that V8 wasn't updated since October 2022, and there was plenty of CVEs and 0-days in V8 since then, so it should be updated to the latest stable release... Note that those CVEs are published with `cpe:2.3:a:google:chrome:*` and not `cpe:2.3:a:google:v8:*`, so perhaps it's worth updating the CVE scanner to account for that, since it would need to map Chrome versions to V8 versions (see: https://omahaproxy.appspot.com/ or the existing instructions about extracting BoringSSL version from Chromium's `DEPS` file). cc @phlax

**Comment by phlax** [comment]
@mpwarres as a first step can we update v8 to the latest stable - given above info we possibly also want to backport the update to the other stable branches

**Comment by htuch** [comment]
@phlax do we have a systematic way to reason about which dependencies have stable main branches and which not for when we pin to non-release?

I agree we can use bazel/v8.patch for now give it is upstreamed. My main point (as always) is that we do not carry indefinitely patches in Envoy that belong upstream. That pattern is not sustainable and hurts velocity in various ways.

**Comment by mpwarres** [comment]
Using bazel/v8.patch for now SGTM. Separately I will work on updating proxy-wasm-cpp-host to the latest stable v8 release, but won't be able to start on that until next week.

**Comment by phlax** [comment]
> do we have a systematic way to reason about which dependencies have stable main branches and which not for when we pin to non-release?

mostly i think trust and necessity - its different for different deps - but its generally either maintainers or codeowners that are responsible for these - or they have no release/some special requirement/etc

i think dep-sheps mostly push back on this if these are being added - i generally take a look to see if there are releases if im reviewing

the cve scanner is far from perfect and some deps are ~invisible to it at least how currently track deps - but non-releases dont help with that - these are also missed by our release/update tracking

probs outside the scope of this quite important ticket tho - maybe we should raise it in a separate issue

**Comment by phlax** [comment]
would be good to get this update prior to next stable - not sure implications of doing that immediately before release

**Comment by mpwarres** [comment]
Sorry, got preempted by other items (and OOO) so not done yet. Working on it now.

**Comment by phlax** [comment]
cc @mpwarres this has a bit more urgency as we are receiving reports of multiple CVEs with current version

cc @nikita96

**Comment by mpwarres** [comment]
Ack, working on it now as my top priority.

**Comment by nikita96** [comment]
Hi @phlax and @mpwarres any plans on upgrading v8 anytime soon? I see a lot of CVEs being reported on current version.
If this is a lower priority item in your list, could you please share more details around how v8 is being used in envoy and how vulnerable is envoy if v8 is not upgraded?

**Comment by phlax** [comment]
i think we need to get this updated asap - ive just raised a PR (#35552) to set the correct CPE - which hopefully should break our CVE scanner

**Comment by phlax** [comment]
hmm - that broke the scanner in unexpected ways - but i dont think this can work

i just read @PiotrSikora 's comment properly about mapping deps - im not sure our scanner can handle that atm - im also not sure how it would be possible to do that without throwing lots of unrelated errors - there is not an obvious way to separate CVEs that are specifically v8 related

**Comment by slimm609** [comment]
Any update on this? Scans are currently showing 31 High and 45 Mediums findings against the V8 version in use.

**Comment by phlax** [comment]
@slimm609 im just looking at this now - it was updated a couple of months ago - are you scanning against the main branch?

the versioning for this is frankly bizarre - but if there are genuinely cves flagging for our current version then would be good to get this updated before we cut the pending release

cc @yanavlasov

**Comment by slimm609** [comment]
ah, i missed the update to 13.6.  I am going to scan and see what remains

**Comment by slimm609** [comment]
https://github.com/envoyproxy/envoy/pull/39581

This is the draft but main still has "10.7.193.13" from 2022 https://github.com/envoyproxy/envoy/blob/main/bazel/repository_locations.bzl#L1182

**Comment by phlax** [comment]
hmm, i see - it wasnt actually updated

the error looks like it should be surmountable

```console

Error: ERROR: /build/bazel_root/base/external/v8/BUILD.bazel:4345:11: no such package '@@v8//third_party/fast_float/src': BUILD file not found in directory 'third_party/fast_float/src' of external repository @@v8. Add a BUILD file to a directory to mark it as a package. and referenced by '@@v8//:v8_libshared_noicu'
```

**Comment by phlax** [comment]
@mpwarres just to flag the urgency on this - we are currently blocking envoy release while this is resolved

im having a go at updating this, and @yanavlasov has also tried recently - but its not trivial, the build system seems to be a mess, and afaict the bazel support is no more than symbolic (perhaps im missing something)

any chance we could get this resolved? its been several years that this has been waiting, and there is a large catalogue of vulnerabilities being flagged

**Comment by mpwarres** [comment]
I have sent out https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/438 and https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/439 to get proxy-wasm-cpp-host building with C++20, which is necessary in order to build more recent versions of V8.

I will try to get out a proxy-wasm-cpp-host PR tomorrow that is stacked on those, to build with the most recent stable V8, 13.8.258.26.

However, I am about to be completely offline until 7/21, so @martijneken or @leonm1 may need to merge the the above PRs into proxy-wasm-cpp-host.

**Comment by mpwarres** [comment]
Ok, here is the chain of PRs that are needed for proxy-wasm-cpp-host, in parent->child order. Once they have been merged into proxy-wasm-cpp-host, updating Envoy to that new version should hopefully work.
* https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/438
* proxy-wasm/proxy-wasm-cpp-host#439 
* proxy-wasm/proxy-wasm-cpp-host#440

I am not going to be available to merge these, since I am about to board a flight after which I will have no online access for a week, but @martijneken or @leonm1 can help with the merging.

**Comment by leonm1** [comment]
As an update, I took over this work while @mpwarres has been out. The PRs changed ownership:

* proxy-wasm/proxy-wasm-cpp-host/pull/438 was merged
* proxy-wasm/proxy-wasm-cpp-host/pull/441 was merged
* proxy-wasm/proxy-wasm-cpp-host/pull/442 is almost ready for merging, just sorting through some CI issues on macOS, whic I don't have available to test locally.

I'll send a PR to Envoy to update to this proxy-wasm and V8 version, initially pointing at my fork to test, then switching to the upstream proxy-wasm/proxy-wasm-cpp-host once the PR is merged.

**Comment by martijneken** [comment]
FWIW we recognize that using a years-old dependency is unacceptable, so we are also working on some process improvements to prevent stale deps in proxy-wasm repos (especially cpp-host) going forward: https://github.com/proxy-wasm/proxy-wasm-cpp-host/issues/443

**Comment by phlax** [comment]
thanks for following up on this - really appreciated

**Comment by leonm1** [comment]
Sent https://github.com/envoyproxy/envoy/pull/40305 to update the V8 + proxy-wasm version. The Proxy-Wasm PR is still working through CI, but I wanted to send the PR once I got it into a tests-passing state.
